// @metaplex-foundation/mpl-candy-machine/dist/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MASTER_EDITION_SIZE = exports.METADATA_SIZE = exports.CANDY_GUARD_DATA = exports.CANDY_GUARD_LABEL_SIZE = exports.CANDY_MACHINE_HIDDEN_SECTION = exports.CONFIG_LINE_SIZE = exports.MAX_CREATOR_LEN = exports.MAX_CREATOR_LIMIT = exports.MAX_URI_LENGTH = exports.MAX_SYMBOL_LENGTH = exports.MAX_NAME_LENGTH = void 0;
exports.MAX_NAME_LENGTH = 32;
exports.MAX_SYMBOL_LENGTH = 10;
exports.MAX_URI_LENGTH = 200;
exports.MAX_CREATOR_LIMIT = 5;
exports.MAX_CREATOR_LEN = 32 + 1 + 1;
exports.CONFIG_LINE_SIZE = 4 + exports.MAX_NAME_LENGTH + 4 + exports.MAX_URI_LENGTH;
exports.CANDY_MACHINE_HIDDEN_SECTION = 8 + // discriminator
    8 + // features
    32 + // authority
    32 + // mint authority
    32 + // collection mint
    8 + // items redeemed
    8 + // items available (config data)
    4 +
    exports.MAX_SYMBOL_LENGTH + // u32 + max symbol length
    2 + // seller fee basis points
    8 + // max supply
    1 + // is mutable
    4 +
    exports.MAX_CREATOR_LIMIT * exports.MAX_CREATOR_LEN + // u32 + creators vec
    1 + // option (config lines settings)
    4 +
    exports.MAX_NAME_LENGTH + // u32 + max name length
    4 + // name length
    4 +
    exports.MAX_URI_LENGTH + // u32 + max uri length
    4 + // uri length
    1 + // is sequential
    1 + // option (hidden setting)
    4 +
    exports.MAX_NAME_LENGTH + // u32 + max name length
    4 +
    exports.MAX_URI_LENGTH + // u32 + max uri length
    32; // hash
exports.CANDY_GUARD_LABEL_SIZE = 6;
exports.CANDY_GUARD_DATA = 8 + // discriminator
    32 + // base
    1 + // bump
    32; // authority
exports.METADATA_SIZE = 679;
exports.MASTER_EDITION_SIZE = 282;
