// @metaplex-foundation/mpl-candy-machine/dist/src/generated/instructions/mintFromCandyMachine.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.mintFromCandyMachine = exports.getMintFromCandyMachineInstructionDataSerializer = void 0;
const mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const shared_1 = require("../shared");
function getMintFromCandyMachineInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })]], { description: 'MintFromCandyMachineInstructionData' }), (value) => ({
        ...value,
        discriminator: [51, 57, 225, 47, 182, 146, 137, 166],
    }));
}
exports.getMintFromCandyMachineInstructionDataSerializer = getMintFromCandyMachineInstructionDataSerializer;
// Instruction.
function mintFromCandyMachine(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCandyMachineCore', 'CndyV3LdqHUfDLmE5naZjVN8rBZz4tqhdefbAnjHG3JR');
    // Accounts.
    const resolvedAccounts = {
        candyMachine: {
            index: 0,
            isWritable: true,
            value: input.candyMachine ?? null,
        },
        authorityPda: {
            index: 1,
            isWritable: true,
            value: input.authorityPda ?? null,
        },
        mintAuthority: {
            index: 2,
            isWritable: false,
            value: input.mintAuthority ?? null,
        },
        payer: { index: 3, isWritable: true, value: input.payer ?? null },
        nftMint: { index: 4, isWritable: true, value: input.nftMint ?? null },
        nftMintAuthority: {
            index: 5,
            isWritable: false,
            value: input.nftMintAuthority ?? null,
        },
        nftMetadata: {
            index: 6,
            isWritable: true,
            value: input.nftMetadata ?? null,
        },
        nftMasterEdition: {
            index: 7,
            isWritable: true,
            value: input.nftMasterEdition ?? null,
        },
        collectionAuthorityRecord: {
            index: 8,
            isWritable: false,
            value: input.collectionAuthorityRecord ?? null,
        },
        collectionMint: {
            index: 9,
            isWritable: false,
            value: input.collectionMint ?? null,
        },
        collectionMetadata: {
            index: 10,
            isWritable: true,
            value: input.collectionMetadata ?? null,
        },
        collectionMasterEdition: {
            index: 11,
            isWritable: false,
            value: input.collectionMasterEdition ?? null,
        },
        collectionUpdateAuthority: {
            index: 12,
            isWritable: false,
            value: input.collectionUpdateAuthority ?? null,
        },
        tokenMetadataProgram: {
            index: 13,
            isWritable: false,
            value: input.tokenMetadataProgram ?? null,
        },
        tokenProgram: {
            index: 14,
            isWritable: false,
            value: input.tokenProgram ?? null,
        },
        systemProgram: {
            index: 15,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        recentSlothashes: {
            index: 16,
            isWritable: false,
            value: input.recentSlothashes ?? null,
        },
    };
    // Default values.
    if (!resolvedAccounts.authorityPda.value) {
        resolvedAccounts.authorityPda.value = (0, hooked_1.findCandyMachineAuthorityPda)(context, { candyMachine: (0, shared_1.expectPublicKey)(resolvedAccounts.candyMachine.value) });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.nftMintAuthority.value) {
        resolvedAccounts.nftMintAuthority.value = context.identity;
    }
    if (!resolvedAccounts.nftMetadata.value) {
        resolvedAccounts.nftMetadata.value = (0, mpl_token_metadata_1.findMetadataPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.nftMint.value),
        });
    }
    if (!resolvedAccounts.nftMasterEdition.value) {
        resolvedAccounts.nftMasterEdition.value = (0, mpl_token_metadata_1.findMasterEditionPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.nftMint.value),
        });
    }
    if (!resolvedAccounts.collectionAuthorityRecord.value) {
        resolvedAccounts.collectionAuthorityRecord.value =
            (0, mpl_token_metadata_1.findCollectionAuthorityRecordPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.collectionMint.value),
                collectionAuthority: (0, shared_1.expectPublicKey)(resolvedAccounts.authorityPda.value),
            });
    }
    if (!resolvedAccounts.collectionMetadata.value) {
        resolvedAccounts.collectionMetadata.value = (0, mpl_token_metadata_1.findMetadataPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.collectionMint.value),
        });
    }
    if (!resolvedAccounts.collectionMasterEdition.value) {
        resolvedAccounts.collectionMasterEdition.value = (0, mpl_token_metadata_1.findMasterEditionPda)(context, { mint: (0, shared_1.expectPublicKey)(resolvedAccounts.collectionMint.value) });
    }
    if (!resolvedAccounts.tokenMetadataProgram.value) {
        resolvedAccounts.tokenMetadataProgram.value = context.programs.getPublicKey('mplTokenMetadata', 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s');
        resolvedAccounts.tokenMetadataProgram.isWritable = false;
    }
    if (!resolvedAccounts.tokenProgram.value) {
        resolvedAccounts.tokenProgram.value = context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
        resolvedAccounts.tokenProgram.isWritable = false;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    if (!resolvedAccounts.recentSlothashes.value) {
        resolvedAccounts.recentSlothashes.value = (0, umi_1.publicKey)('SysvarS1otHashes111111111111111111111111111');
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getMintFromCandyMachineInstructionDataSerializer().serialize({});
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.mintFromCandyMachine = mintFromCandyMachine;
