// @metaplex-foundation/mpl-candy-machine/dist/src/generated/instructions/route.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.route = exports.getRouteInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getRouteInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })],
        ['guard', (0, types_1.getGuardTypeSerializer)()],
        ['data', (0, serializers_1.bytes)({ size: (0, serializers_1.u32)() })],
        ['group', (0, serializers_1.option)((0, serializers_1.string)())],
    ], { description: 'RouteInstructionData' }), (value) => ({
        ...value,
        discriminator: [229, 23, 203, 151, 122, 227, 173, 42],
    }));
}
exports.getRouteInstructionDataSerializer = getRouteInstructionDataSerializer;
// Instruction.
function route(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCandyGuard', 'Guard1JwRhJkVH6XZhzoYxeBVQe872VH6QggF4BWmS9g');
    // Accounts.
    const resolvedAccounts = {
        candyGuard: {
            index: 0,
            isWritable: false,
            value: input.candyGuard ?? null,
        },
        candyMachine: {
            index: 1,
            isWritable: true,
            value: input.candyMachine ?? null,
        },
        payer: { index: 2, isWritable: true, value: input.payer ?? null },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.candyGuard.value) {
        resolvedAccounts.candyGuard.value = (0, hooked_1.findCandyGuardPda)(context, {
            base: (0, shared_1.expectPublicKey)(resolvedAccounts.candyMachine.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getRouteInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.route = route;
