// @metaplex-foundation/mpl-candy-machine/dist/src/generated/instructions/setCollection.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setCollection = exports.getSetCollectionInstructionDataSerializer = void 0;
const mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const shared_1 = require("../shared");
function getSetCollectionInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })]], { description: 'SetCollectionInstructionData' }), (value) => ({
        ...value,
        discriminator: [192, 254, 206, 76, 168, 182, 59, 223],
    }));
}
exports.getSetCollectionInstructionDataSerializer = getSetCollectionInstructionDataSerializer;
// Instruction.
function setCollection(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCandyMachineCore', 'CndyV3LdqHUfDLmE5naZjVN8rBZz4tqhdefbAnjHG3JR');
    // Accounts.
    const resolvedAccounts = {
        candyMachine: {
            index: 0,
            isWritable: true,
            value: input.candyMachine ?? null,
        },
        authority: { index: 1, isWritable: false, value: input.authority ?? null },
        authorityPda: {
            index: 2,
            isWritable: true,
            value: input.authorityPda ?? null,
        },
        payer: { index: 3, isWritable: false, value: input.payer ?? null },
        collectionMint: {
            index: 4,
            isWritable: false,
            value: input.collectionMint ?? null,
        },
        collectionMetadata: {
            index: 5,
            isWritable: false,
            value: input.collectionMetadata ?? null,
        },
        collectionAuthorityRecord: {
            index: 6,
            isWritable: true,
            value: input.collectionAuthorityRecord ?? null,
        },
        newCollectionUpdateAuthority: {
            index: 7,
            isWritable: true,
            value: input.newCollectionUpdateAuthority ?? null,
        },
        newCollectionMetadata: {
            index: 8,
            isWritable: false,
            value: input.newCollectionMetadata ?? null,
        },
        newCollectionMint: {
            index: 9,
            isWritable: false,
            value: input.newCollectionMint ?? null,
        },
        newCollectionMasterEdition: {
            index: 10,
            isWritable: false,
            value: input.newCollectionMasterEdition ?? null,
        },
        newCollectionAuthorityRecord: {
            index: 11,
            isWritable: true,
            value: input.newCollectionAuthorityRecord ?? null,
        },
        tokenMetadataProgram: {
            index: 12,
            isWritable: false,
            value: input.tokenMetadataProgram ?? null,
        },
        systemProgram: {
            index: 13,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Default values.
    if (!resolvedAccounts.authority.value) {
        resolvedAccounts.authority.value = context.identity;
    }
    if (!resolvedAccounts.authorityPda.value) {
        resolvedAccounts.authorityPda.value = (0, hooked_1.findCandyMachineAuthorityPda)(context, { candyMachine: (0, shared_1.expectPublicKey)(resolvedAccounts.candyMachine.value) });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.collectionMetadata.value) {
        resolvedAccounts.collectionMetadata.value = (0, mpl_token_metadata_1.findMetadataPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.collectionMint.value),
        });
    }
    if (!resolvedAccounts.collectionAuthorityRecord.value) {
        resolvedAccounts.collectionAuthorityRecord.value =
            (0, mpl_token_metadata_1.findCollectionAuthorityRecordPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.collectionMint.value),
                collectionAuthority: (0, shared_1.expectPublicKey)(resolvedAccounts.authorityPda.value),
            });
    }
    if (!resolvedAccounts.newCollectionMetadata.value) {
        resolvedAccounts.newCollectionMetadata.value = (0, mpl_token_metadata_1.findMetadataPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.newCollectionMint.value),
        });
    }
    if (!resolvedAccounts.newCollectionMasterEdition.value) {
        resolvedAccounts.newCollectionMasterEdition.value = (0, mpl_token_metadata_1.findMasterEditionPda)(context, { mint: (0, shared_1.expectPublicKey)(resolvedAccounts.newCollectionMint.value) });
    }
    if (!resolvedAccounts.newCollectionAuthorityRecord.value) {
        resolvedAccounts.newCollectionAuthorityRecord.value =
            (0, mpl_token_metadata_1.findCollectionAuthorityRecordPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.newCollectionMint.value),
                collectionAuthority: (0, shared_1.expectPublicKey)(resolvedAccounts.authorityPda.value),
            });
    }
    if (!resolvedAccounts.tokenMetadataProgram.value) {
        resolvedAccounts.tokenMetadataProgram.value = context.programs.getPublicKey('mplTokenMetadata', 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s');
        resolvedAccounts.tokenMetadataProgram.isWritable = false;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getSetCollectionInstructionDataSerializer().serialize({});
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.setCollection = setCollection;
