// @metaplex-foundation/mpl-candy-machine/dist/src/generated/instructions/setTokenStandard.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setTokenStandard = exports.getSetTokenStandardInstructionDataSerializer = void 0;
const mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const shared_1 = require("../shared");
function getSetTokenStandardInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })],
        ['tokenStandard', (0, mpl_token_metadata_1.getTokenStandardSerializer)()],
    ], { description: 'SetTokenStandardInstructionData' }), (value) => ({
        ...value,
        discriminator: [147, 212, 106, 195, 30, 170, 209, 128],
    }));
}
exports.getSetTokenStandardInstructionDataSerializer = getSetTokenStandardInstructionDataSerializer;
// Instruction.
function setTokenStandard(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCandyMachineCore', 'CndyV3LdqHUfDLmE5naZjVN8rBZz4tqhdefbAnjHG3JR');
    // Accounts.
    const resolvedAccounts = {
        candyMachine: {
            index: 0,
            isWritable: true,
            value: input.candyMachine ?? null,
        },
        authority: { index: 1, isWritable: false, value: input.authority ?? null },
        authorityPda: {
            index: 2,
            isWritable: true,
            value: input.authorityPda ?? null,
        },
        payer: { index: 3, isWritable: true, value: input.payer ?? null },
        ruleSet: { index: 4, isWritable: false, value: input.ruleSet ?? null },
        collectionDelegateRecord: {
            index: 5,
            isWritable: true,
            value: input.collectionDelegateRecord ?? null,
        },
        collectionMint: {
            index: 6,
            isWritable: false,
            value: input.collectionMint ?? null,
        },
        collectionMetadata: {
            index: 7,
            isWritable: true,
            value: input.collectionMetadata ?? null,
        },
        collectionAuthorityRecord: {
            index: 8,
            isWritable: true,
            value: input.collectionAuthorityRecord ?? null,
        },
        collectionUpdateAuthority: {
            index: 9,
            isWritable: false,
            value: input.collectionUpdateAuthority ?? null,
        },
        tokenMetadataProgram: {
            index: 10,
            isWritable: false,
            value: input.tokenMetadataProgram ?? null,
        },
        systemProgram: {
            index: 11,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        sysvarInstructions: {
            index: 12,
            isWritable: false,
            value: input.sysvarInstructions ?? null,
        },
        authorizationRulesProgram: {
            index: 13,
            isWritable: false,
            value: input.authorizationRulesProgram ?? null,
        },
        authorizationRules: {
            index: 14,
            isWritable: false,
            value: input.authorizationRules ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.authority.value) {
        resolvedAccounts.authority.value = context.identity;
    }
    if (!resolvedAccounts.authorityPda.value) {
        resolvedAccounts.authorityPda.value = (0, hooked_1.findCandyMachineAuthorityPda)(context, { candyMachine: (0, shared_1.expectPublicKey)(resolvedAccounts.candyMachine.value) });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.collectionDelegateRecord.value) {
        resolvedAccounts.collectionDelegateRecord.value =
            (0, mpl_token_metadata_1.findMetadataDelegateRecordPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.collectionMint.value),
                delegateRole: mpl_token_metadata_1.MetadataDelegateRole.Collection,
                updateAuthority: (0, shared_1.expectPublicKey)(resolvedAccounts.collectionUpdateAuthority.value),
                delegate: (0, shared_1.expectPublicKey)(resolvedAccounts.authorityPda.value),
            });
    }
    if (!resolvedAccounts.collectionMetadata.value) {
        resolvedAccounts.collectionMetadata.value = (0, mpl_token_metadata_1.findMetadataPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.collectionMint.value),
        });
    }
    if (!resolvedAccounts.tokenMetadataProgram.value) {
        resolvedAccounts.tokenMetadataProgram.value = context.programs.getPublicKey('mplTokenMetadata', 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s');
        resolvedAccounts.tokenMetadataProgram.isWritable = false;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    if (!resolvedAccounts.sysvarInstructions.value) {
        resolvedAccounts.sysvarInstructions.value = (0, umi_1.publicKey)('Sysvar1nstructions1111111111111111111111111');
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getSetTokenStandardInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.setTokenStandard = setTokenStandard;
