// @metaplex-foundation/mpl-candy-machine/dist/src/generated/programs/mplCandyGuard.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMplCandyGuardProgramId = exports.getMplCandyGuardProgram = exports.createMplCandyGuardProgram = exports.MPL_CANDY_GUARD_PROGRAM_ID = void 0;
const errors_1 = require("../errors");
exports.MPL_CANDY_GUARD_PROGRAM_ID = 'Guard1JwRhJkVH6XZhzoYxeBVQe872VH6QggF4BWmS9g';
function createMplCandyGuardProgram() {
    return {
        name: 'mplCandyGuard',
        publicKey: exports.MPL_CANDY_GUARD_PROGRAM_ID,
        getErrorFromCode(code, cause) {
            return (0, errors_1.getMplCandyGuardErrorFromCode)(code, this, cause);
        },
        getErrorFromName(name, cause) {
            return (0, errors_1.getMplCandyGuardErrorFromName)(name, this, cause);
        },
        isOnCluster() {
            return true;
        },
    };
}
exports.createMplCandyGuardProgram = createMplCandyGuardProgram;
function getMplCandyGuardProgram(context, clusterFilter) {
    return context.programs.get('mplCandyGuard', clusterFilter);
}
exports.getMplCandyGuardProgram = getMplCandyGuardProgram;
function getMplCandyGuardProgramId(context, clusterFilter) {
    return context.programs.getPublicKey('mplCandyGuard', exports.MPL_CANDY_GUARD_PROGRAM_ID, clusterFilter);
}
exports.getMplCandyGuardProgramId = getMplCandyGuardProgramId;
