// @metaplex-foundation/mpl-candy-machine/dist/src/generated/types/freezeInstruction.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFreezeInstructionSerializer = exports.FreezeInstruction = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var FreezeInstruction;
(function (FreezeInstruction) {
    FreezeInstruction[FreezeInstruction["Initialize"] = 0] = "Initialize";
    FreezeInstruction[FreezeInstruction["Thaw"] = 1] = "Thaw";
    FreezeInstruction[FreezeInstruction["UnlockFunds"] = 2] = "UnlockFunds";
})(FreezeInstruction = exports.FreezeInstruction || (exports.FreezeInstruction = {}));
function getFreezeInstructionSerializer() {
    return (0, serializers_1.scalarEnum)(FreezeInstruction, {
        description: 'FreezeInstruction',
    });
}
exports.getFreezeInstructionSerializer = getFreezeInstructionSerializer;
